pkg/build/kube/tarbits.go

```diff
@@ -0,0 +1,425 @@
+	paths := map[string]string{
+		// docker images
+		filepath.Join(b.tarRoot, "kube-apiserver.tar"):          "images/kube-apiserver.tar",
+		filepath.Join(b.tarRoot, "kube-controller-manager.tar"): "images/kube-controller-manager.tar",
+		filepath.Join(b.tarRoot, "kube-scheduler.tar"):          "images/kube-scheduler.tar",
+		filepath.Join(b.tarRoot, "kube-proxy.tar"):              "images/kube-proxy.tar",
+		// version file
+		filepath.Join(b.tarRoot, "version"): "version",
+		// kubelet service
+		filepath.Join(b.tarRoot, "kubelet.service"): "systemd/kubelet.service",
+		// kubeadm config
+		filepath.Join(b.tarRoot, "10-kubeadm.conf"): "systemd/10-kubeadm.conf",
+		// binaries
+		filepath.Join(b.tarRoot, "kubeadm"): "bin/kubeadm",
+		filepath.Join(b.tarRoot, "kubectl"): "bin/kubectl",
+		filepath.Join(b.tarRoot, "kubelet"): "bin/kubelet",
+	}
+
+	return paths
+}
+
+func (b *TarBits) verifyPaths() error {
+	// Verify the files exist and ensure the binaries' permissions
+	// include the executable bit.
+	for k, v := range b.paths {
+		if _, err := os.Stat(k); err != nil {
+			if os.IsNotExist(err) {
+				switch v {
+				case "systemd/kubelet.service":
+					err := ioutil.WriteFile(k, []byte(kubeletService), 0644)
+					if err != nil {
+						return err
+					}
+				case "systemd/10-kubeadm.conf":
+					err := ioutil.WriteFile(k, []byte(tenKubeadmConf), 0644)
+					if err != nil {
+						return err
+					}
+				default:
+					return err
+				}
+			} else {
+				return err
+			}
+		}
+		if strings.HasPrefix(v, "bin/") {
+			os.Chmod(k, 0755)
+		}
+	}
+	return nil
+}
+
+func (b *TarBits) getRealURIForCIOrRelease() error {
+	resp, err := http.Head(b.remoteURI + "kubernetes.tar.gz")
+	if err == nil && resp.StatusCode == http.StatusOK {
+		return nil
+	}
+
+	_, r, err := b.getReaderForURI(b.remoteURI)
+	if err != nil {
+		b.remoteURI = b.remoteURI + ".txt"
+		_, r, err = b.getReaderForURI(b.remoteURI)
+	}
+	if err != nil {
+		return nil
+	}
+	defer r.Close()
+
+	buf, err := ioutil.ReadAll(r)
+	if err != nil {
+		return err
+	}
+
+	version := url.PathEscape(string(bytes.TrimSpace(buf)))
+
+	if b.tbitsType == tarBitsCI {
+		b.remoteURI = fmt.Sprintf("%s/ci/%s", ciURIPrefix, version)
+	} else {
+		b.remoteURI = fmt.Sprintf("%s/release/%s", releaseURIPrefix, version)
+	}
+
+	return nil
+}
+
+func (b *TarBits) getReaderForURI(uri string) (int64, io.ReadCloser, error) {
+	resp, err := http.Get(uri)
+	if err != nil {
+		return 0, nil, fmt.Errorf("failed to read %s: %v", uri, err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		return 0, nil, fmt.Errorf("%s: %s", uri, resp.Status)
+	}
+	return resp.ContentLength, resp.Body, nil
+}
+
+func (b *TarBits) copyFromURI(src, dst string) error {
+	n, r, err := b.getReaderForURI(src)
+	if err != nil {
+		return err
+	}
+	defer r.Close()
+
+	// If the file already exists and has the same size as the remote
+	// content then do not redownload it.
+	if f, err := os.Stat(dst); err == nil {
```

Comment:
needs `os.IsNotExist` here

Reply:
Hi @neolit123,

So I'm aware of the function, but I explicitly opted not to use it because in this case I don't care *why* the `stat` failed. If the `stat` fails I want to try and create and write to the file. If the write fails, then an error will be returned. It's a case of preferring `create -> success|fail` versus `check -> create -> success|fail`. If the `check` step fails for any reason then `create` should at least be attempted. Therefore I'm just optimizing away the initial `check` step.

Reply:
Hi @neolit123,

By the way, using `os.IsNotExist` here is also unnecessary since I only care if it exists, not if it doesn't. It's a fine distinction, but `os.Stat` returns a `nil` error if a file exists. The only time you need `os.IsNotExist` is when `os.Stat` *does* return an error.

It's a common misconception that `os.IsExist` is the opposite of `os.IsNotExist` when that's not accurate.The `os.IsExist` function is for when a file operation fails because the operation is attempting to modify an existing file in a disallowed manner. For example, if `os.Symlink(oldname, newname string)` is invoked and the given `newname` points to an existing file or directory, an error will be returned such that `os.IsExist(err) == true`.

The `os.IsNotExistFor` function is for determining whether or not the reason the `os.Stat` or any other operation that accesses a file fails because the file doesn't actually exist -- not whether or not the process can access the file. For example, if the executing process lacks permission to access the file then `os.Stat` returns an error such that `os.IsPermission(err) == true`. In this case `os.IsNotExist(err) == false`, but the process is still unable to stat the file.